clients/src/test/java/org/apache/kafka/test/TestUtils.java

```diff
@@ -142,26 +142,40 @@ public static String randomString(final int len) {
     }
 
     /**
-     * Create an empty file in the default temporary-file directory, using `kafka` as the prefix and `tmp` as the
-     * suffix to generate its name.
+     * Create an empty file in the default temporary-file directory, using the given prefix and suffix
+     * to generate its name.
+     * @throws IOException
      */
-    public static File tempFile() throws IOException {
-        final File file = File.createTempFile("kafka", ".tmp");
+    public static File tempFile(final String prefix, final String suffix) throws IOException {
+        final File file = Files.createTempFile(prefix, suffix).toFile();
         file.deleteOnExit();
 
+        Exit.addShutdownHook("delete-temp-file-shutdown-hook", () -> {
```

Comment:
@divijvaidya I've come across this while doing research for an unrelated PR; sorry to comment so late.

Do you know why we use `Exit::addShutdownHook` here instead of invoking `Runtime.getRuntime().addShutdownHook` directly? If we've overridden the shutdown hook behavior in the `Exit` class during testing (which we don't do anywhere in the code base ATM, but may do at some point in the future), then won't this shutdown hook be potentially disabled?

Reply:
From what I understand, we try to use `Exit.scala` instead of `Runtime.getRuntime()` so that Tests may provide their own implementation of shutdown hook adder. This is mentioned here: https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/utils/Exit.scala#L22 

I didn't understand the scenario where this hook will be disabled! Could you please explain with an example?

If a test would override the shutdown hook behaviour, then I would assume it will add a new shutdown hook? or perhaps, add a new `ShutdownHookAdder` where it may play with the order of the hooks? In all such scenarios, this cleanup hook will still be executed as long it is registered correctly.

Reply:
The idea (at least with calls to terminate the JVM) is that we can add a layer in between the components under test and the exit API provided by the SDK, and then alter that layer during testing to prevent the actual API from being accessed.

One example is in the embedded Connect clusters we use for our integration tests: https://github.com/apache/kafka/blob/dbf5826cd545820652ed6d136727aa80a03f4f54/connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectCluster.java#L111-L120

With shutdown hooks, it's mostly hypothetical at this point since AFAICT there are no actual invocations of `Exit::setShutdownHookAdder` in the code base, but I believe the intention would be similar. You'd intercept attempts by components under test to add shutdown hooks, both to prevent those hooks from actually being added (in other words, to prevent `Runtime.getRuntime().addShutdownHook` from being called), and possibly to make assertions on those hooks.

We don't want to allow for the shutdown hooks we added in this PR to be intercepted, right? If so, we might consider invoking `Runtime.getRuntime().addShutdownHook` directly here.

Reply:
@C0urante Makes sense. We should be using `Runtime.getRuntime().addShutdownHook` here because we don't want the Exit interceptors added via the tests to override this operation of cleaning up the files. I will file a PR to fix it.